apps/src/commands.js

```diff
@@ -0,0 +1,222 @@
+ * operations that need to take place with a particular relationship to time
+ * or to one another - asynchronous calls that should happen in order, events
+ * that should be triggered on a timeline, etc.  Instead of nesting callbacks
+ * N-layers-deep, create a queue of commands to run.
+ *
+ * You'll usually want to define commands for your own unique needs.  A child
+ * command should always call the Command constructor from its own constructor,
+ * it will almost always need to override onBegin_, and it may want to override
+ * tick and onEnd_.  Commands can call success() or fail() on themselves, or
+ * wait for others to do so; usage is up to you.
+ *
+ * @constructor
+ */
+var Command = commands.Command = function () {
+  /**
+   * @type {commandState}
+   * @private
+   */
+  this.status_ = commandState.NOT_STARTED;
+};
+
+/**
+ * Whether the command has started working.
+ * @returns {boolean}
+ */
+Command.prototype.isStarted = function () {
+  return this.status_ !== commandState.NOT_STARTED;
+};
+
+/**
+ * Whether the command has succeeded or failed, and is
+ * finished with its work.
+ * @returns {boolean}
+ */
+Command.prototype.isFinished = function () {
+  return this.succeeded() || this.failed();
+};
+
+/**
+ * Whether the command has finished with its work and reported success.
+ * @returns {boolean}
+ */
+Command.prototype.succeeded = function () {
+  return this.status_ === commandState.SUCCESS;
+};
+
+/**
+ * Whether the command has finished with its work and reported failure.
+ * @returns {boolean}
+ */
+Command.prototype.failed = function () {
+  return this.status_ === commandState.FAILURE;
+};
+
+/**
+ * Tells the command to start working.
+ */
+Command.prototype.begin = function () {
+  this.status_ = commandState.WORKING;
+  this.onBegin_();
+};
+
+/**
+ * Called to mark that the command is done with its work and has
+ * succeeded.  Might be called by the command itself, or by an external
+ * controller.
+ * @throws if called before begin()
+ */
+Command.prototype.succeed = function () {
+  if (!this.isStarted()) {
+    throw new Error("Command cannot succeed before it begins.");
+  }
+  this.status_ = commandState.SUCCESS;
+  this.onEnd_();
+};
+
+/**
+ * Called to mark that the command is done with its work and has
+ * failed.  Might be called by the command itself, or by an external
+ * controller.
+ * @throws if called before begin()
+ */
+Command.prototype.fail = function () {
+  if (!this.isStarted()) {
+    throw new Error("Command cannot fail before it begins.");
+  }
+  this.status_ = commandState.FAILURE;
+  this.onEnd_();
+};
+
+/**
+ * Stub to be implemented by descendant classes, of operations to perform
+ * when the command begins.
+ * @private
+ */
+Command.prototype.onBegin_ = function () {};
+
+/**
+ * Stub to be implemented by descendant classes, of operations to perform
+ * on tick.
+ * @param {RunLoop.Clock} clock - Time information passed into all tick methods.
```

Comment:
Ah, no, RunLoop is not in top level.  It could be, it's not NetSim-specific code.